executor/src/subscriber.rs

```diff
@@ -1,112 +1,148 @@
 
 /// The `Subscriber` receives certificates sequenced by the consensus and waits until the
 /// downloaded all the transactions references by the certificates; it then
 /// forward the certificates to the Executor Core.
-pub struct Subscriber {
-    /// The temporary storage holding all transactions' data (that may be too big to hold in memory).
-    store: Store<(CertificateDigest, BatchDigest), Batch>,
+pub struct Subscriber<Network> {
     /// Receive reconfiguration updates.
     rx_reconfigure: watch::Receiver<ReconfigureNotification>,
     /// A channel to receive consensus messages.
     rx_consensus: metered_channel::Receiver<ConsensusOutput>,
-    /// A channel to send the complete and ordered list of consensus outputs to the executor. This
-    /// channel is used once all transactions data are downloaded.
-    tx_executor: metered_channel::Sender<ConsensusOutput>,
-    // A channel to send commands to the block waiter to receive
-    // a certificate's batches (block).
-    tx_get_block_commands: metered_channel::Sender<BlockCommand>,
-    // When asking for a certificate's payload we want to retry until we succeed, unless
-    // some irrecoverable error occurs. For that reason a backoff policy is defined
-    get_block_retry_policy: ExponentialBackoff,
+    /// Ordered batches for the consumer
+    tx_notifier: metered_channel::Sender<(BatchIndex, Batch)>,
     /// The metrics handler
     metrics: Arc<ExecutorMetrics>,
+    fetcher: Fetcher<Network>,
 }
 
-impl Subscriber {
-    /// Spawn a new subscriber in a new tokio task.
-    #[must_use]
-    pub fn spawn(
-        store: Store<(CertificateDigest, BatchDigest), Batch>,
-        tx_get_block_commands: metered_channel::Sender<BlockCommand>,
-        rx_reconfigure: watch::Receiver<ReconfigureNotification>,
-        rx_consensus: metered_channel::Receiver<ConsensusOutput>,
-        tx_executor: metered_channel::Sender<ConsensusOutput>,
-        metrics: Arc<ExecutorMetrics>,
-        restored_consensus_output: Vec<ConsensusOutput>,
-    ) -> JoinHandle<()> {
-        let get_block_retry_policy = ExponentialBackoff {
-            initial_interval: Duration::from_millis(500),
-            randomization_factor: backoff::default::RANDOMIZATION_FACTOR,
-            multiplier: backoff::default::MULTIPLIER,
-            max_interval: Duration::from_secs(10), // Maximum backoff is 10 seconds
-            max_elapsed_time: None, // Never end retrying unless a non recoverable error occurs.
-            ..Default::default()
-        };
+struct Fetcher<Network> {
+    network: Network,
+    metrics: Arc<ExecutorMetrics>,
+}
 
-        tokio::spawn(async move {
-            Self {
-                store,
-                rx_reconfigure,
-                rx_consensus,
-                tx_executor,
-                tx_get_block_commands,
-                get_block_retry_policy,
-                metrics,
-            }
+#[must_use]
+pub fn spawn_subscriber(
```

Comment:
@andll are you going to add any test our the subscriber it self? I think it would be great if we could have some unit test coverage for that as well to ensure we don't accidentally break the behaviour